app/validators/rules/rule_validator.py

```diff
@@ -0,0 +1,215 @@
+from app.validators.validator import Validator
+
+
+class Operator:
+    NOT = "not"
+    AND = "and"
+    OR = "or"
+    EQUAL = "=="
+    NOT_EQUAL = "!="
+    GREATER_THAN = ">"
+    LESS_THAN = "<"
+    GREATER_THAN_OR_EQUAL = ">="
+    LESS_THAN_OR_EQUAL = "<="
+    IN = "in"
+    ALL_IN = "all-in"
+    ANY_IN = "any-in"
+    COUNT = "count"
+    DATE = "date"
+    FORMAT_DATE = "format-date"
+    DATE_RANGE = "date-range"
+    MAP = "map"
+
+
+LOGIC_OPERATORS = [Operator.NOT, Operator.AND, Operator.OR]
+
+COMPARISON_OPERATORS = [
+    Operator.EQUAL,
+    Operator.NOT_EQUAL,
+    Operator.GREATER_THAN,
+    Operator.LESS_THAN,
+    Operator.GREATER_THAN_OR_EQUAL,
+    Operator.LESS_THAN_OR_EQUAL,
+]
+
+ARRAY_OPERATORS = [Operator.IN, Operator.ALL_IN, Operator.ANY_IN]
+
+VALUE_OPERATORS = [
+    Operator.DATE,
+    Operator.COUNT,
+    Operator.FORMAT_DATE,
+    Operator.DATE_RANGE,
+    Operator.MAP,
+]
+
+ALL_OPERATORS = (
+    LOGIC_OPERATORS + COMPARISON_OPERATORS + ARRAY_OPERATORS + VALUE_OPERATORS
+)
+
+
+SELF_REFERENCE_KEY = "self"
+
+
+class RulesValidator(Validator):
+    VALUE_DOESNT_EXIST_IN_ANSWER_OPTIONS = "Value doesn't exist in answer options"
+    DATE_OPERATOR_REFERENCES_NON_DATE_ANSWER = (
+        "Date operator references non Date, MonthYearDate, or YearDate answer"
+    )
+    COUNT_OPERATOR_REFERENCES_NON_CHECKBOX_ANSWER = (
+        "Count operator references non Checkbox answer"
+    )
+    MAP_OPERATOR_WITHOUT_SELF_REFERENCE = (
+        f"Argument one of the `map` operator does not reference `{SELF_REFERENCE_KEY}`"
+    )
+    SELF_REFERENCE_OUTSIDE_MAP_OPERATOR = (
+        f"Reference to {SELF_REFERENCE_KEY} was made outside of the `map` operator"
+    )
+
+    def __init__(self, rules, origin_id, questionnaire_schema):
```

Comment:
Is it worth adding type hints to this new/retouched code? Or was that just something for runner?

Reply:
Ya, we don't use `mypy` on this repo. Not as useful here atm, as we got bigger things to prob consider for validator such as using AJV/upgrading Json schema drafts etc.